File: runtime-templates/nodejs/vajra_runtime.js
const express = require('express');
const { Logging } = require('@google-cloud/logging');
const fs = require('fs');
const path = require('path');

const app = express();
app.use(express.json());

// Initialize logging
const logging = new Logging();
const logger = logging.log('vajra-function');

let userFunction = null;

// Load user function
function loadUserFunction() {
    try {
        const handler = process.env.FUNCTION_TARGET || 'handler';
        const mainPath = path.join(__dirname, 'index.js');
        
        if (fs.existsSync(mainPath)) {
            delete require.cache[require.resolve('./index.js')];
            const mainModule = require('./index.js');
            
            if (typeof mainModule[handler] === 'function') {
                userFunction = mainModule[handler];
                logger.info('Function loaded successfully');
            } else {
                throw new Error(`Handler '${handler}' not found in index.js`);
            }
        } else {
            throw new Error('index.js not found');
        }
    } catch (error) {
        logger.error(`Failed to load function: ${error.message}`);
        throw error;
    }
}

// Load function at startup
try {
    loadUserFunction();
} catch (error) {
    console.error('Startup failed:', error.message);
}

app.all('/', async (req, res) => {
    if (!userFunction) {
        return res.status(500).json({ error: 'Function not loaded' });
    }
    
    const startTime = Date.now();
    
    try {
        // Get request data
        const data = req.method === 'POST' ? req.body : req.query;
        
        // Log invocation
        logger.info('Function invoked', {
            method: req.method,
            payloadSize: JSON.stringify(data).length
        });
        
        // Execute user function
        const result = await userFunction(data);
        
        const executionTime = Date.now() - startTime;
        
        // Log success
        logger.info('Function executed successfully', {
            executionTimeMs: executionTime
        });
        
        res.json({
            result: result,
            executionTime: `${executionTime}ms`
        });
        
    } catch (error) {
        const executionTime = Date.now() - startTime;
        
        // Log error
        logger.error('Function execution failed', {
            error: error.message,
            executionTimeMs: executionTime,
            stack: error.stack
        });
        
        res.status(500).json({
            error: error.message,
            executionTime: `${executionTime}ms`
        });
    }
});

app.get('/health', (req, res) => {
    res.json({
        status: 'healthy',
        functionLoaded: userFunction !== null
    });
});

const port = process.env.PORT || 8080;
app.listen(port, '0.0.0.0', () => {
    console.log(`Vajra Node.js runtime listening on port ${port}`);
});

File: runtime-templates/python/vajra_runtime.py
import os
import sys
import json
import importlib.util
import traceback
from flask import Flask, request, jsonify
from google.cloud import logging as cloud_logging

app = Flask(__name__)

# Initialize logging
logging_client = cloud_logging.Client()
logger = logging_client.logger("vajra-function")

def load_user_function():
    """Load user function dynamically"""
    handler = os.environ.get('FUNCTION_TARGET', 'main')
    
    try:
        # Import main module
        if os.path.exists('main.py'):
            spec = importlib.util.spec_from_file_location("main", "main.py")
            main_module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(main_module)
            
            # Get handler function
            if hasattr(main_module, handler):
                return getattr(main_module, handler)
            else:
                raise AttributeError(f"Handler '{handler}' not found in main.py")
        else:
            raise FileNotFoundError("main.py not found")
            
    except Exception as e:
        logger.log_struct({
            "message": f"Failed to load function: {str(e)}",
            "level": "ERROR",
            "traceback": traceback.format_exc()
        })
        raise

# Load function at startup
try:
    user_function = load_user_function()
    logger.log_struct({"message": "Function loaded successfully", "level": "INFO"})
except Exception as e:
    logger.log_struct({"message": f"Startup failed: {str(e)}", "level": "ERROR"})
    user_function = None

@app.route('/', methods=['POST', 'GET'])
def execute_function():
    if not user_function:
        return jsonify({"error": "Function not loaded"}), 500
    
    start_time = time.time()
    
    try:
        # Get request data
        if request.method == 'POST':
            data = request.get_json() or {}
        else:
            data = dict(request.args)
        
        # Log invocation
        logger.log_struct({
            "message": "Function invoked",
            "method": request.method,
            "payload_size": len(json.dumps(data))
        })
        
        # Execute user function
        result = user_function(data)
        
        execution_time = (time.time() - start_time) * 1000
        
        # Log success
        logger.log_struct({
            "message": "Function executed successfully",
            "execution_time_ms": execution_time,
            "level": "INFO"
        })
        
        return jsonify({
            "result": result,
            "execution_time": f"{execution_time:.2f}ms"
        })
        
    except Exception as e:
        execution_time = (time.time() - start_time) * 1000
        error_msg = str(e)
        
        # Log error
        logger.log_struct({
            "message": f"Function execution failed: {error_msg}",
            "execution_time_ms": execution_time,
            "level": "ERROR",
            "traceback": traceback.format_exc()
        })
        
        return jsonify({
            "error": error_msg,
            "execution_time": f"{execution_time:.2f}ms"
        }), 500

@app.route('/health')
def health_check():
    return jsonify({"status": "healthy", "function_loaded": user_function is not None})

if __name__ == '__main__':
    import time
    port = int(os.environ.get('PORT', 8080))
    app.run(host='0.0.0.0', port=port)